tweet_recommendations/other_methods/dbscan_based_method.py

```diff
@@ -0,0 +1,94 @@
+from collections import namedtuple
+from itertools import chain
+from typing import *
+
+import numpy as np
+import pandas as pd
+from scipy.spatial.distance import cosine
+from sklearn.cluster import DBSCAN
+from sklearn.neighbors.kd_tree import KDTree
+
+from ..estimator import Estimator
+
+CentroidData = namedtuple("CentroidData", ["cluster_medoid", "popular_hashtags"])
+
+
+class DBScanBasedEstimator(Estimator):
+    def __init__(self,
+                 verbose: bool = False):
+        self._clusters: np.ndarray = None
+
+        self.clusterizer = None
+        self.neighbours = None
+        self.centroids_data = {}
+
+        self.verbose = verbose
+
+    def fit(self, x: pd.DataFrame, y: Optional[pd.DataFrame] = None, **fit_params):
+        """
+        Builds hashtags representations using w2v method and clusters them using DBScan.
+        :param x: pd.DataFrame with tweet content, user id, and separate hashtags.
+                It is "original_tweets.p" in our case.
+        :param y: None, needed for compatibility.
+        :param fit_params:
+            minimal_hashtag_occurrence: int. If hashtag occurred less than this number
+            then it's not considered during prediction (simply removed).
+            To include all hashtags put number <= 0.
+        :return: self.
+        """
+        minimal_hashtag_occurence = fit_params["minimal_hashtag_occurence"]
+
+        x = self.drop_tweets_with_hashtags_that_occurred_less_than(x, minimal_hashtag_occurence)
+        embeddings = np.array(x["embedding"].to_list())
+
+        print(embeddings.shape)
+
+        self.neighbours = KDTree(embeddings, metric="euclidean")
+        epsilon = np.mean(self.neighbours.query(embeddings, k=2)[0][:, 1])
+        print(f"Found epsilon: {epsilon}")
+
+        self.clusterizer = DBSCAN(metric='manhattan', eps=epsilon, min_samples=1)
+        x["cluster_label"] = self.clusterizer.fit_predict(embeddings)
+
+        print("Clustering finished.")
+
+        x["hashtags"] = x["hashtags"].apply(lambda r: [elem["text"] for elem in r])
+
+        grouped_clusters = x.groupby(["cluster_label"])
+
+        print("Clusters grouped. Building cluster data")
+
+        for cluster_number, cluster in grouped_clusters:
+            cluster_embeddings = np.array(cluster["embedding"].to_list())
+            cluster_center = np.average(cluster_embeddings, axis=0).reshape(1, -1)
+            kd_tree = KDTree(cluster_embeddings, metric="euclidean")
+            cluster_medoid_index = kd_tree.query(cluster_center, k=1)[1][0]
+            cluster_medoid = cluster_embeddings[cluster_medoid_index]
+            aggregated_hashtags = list(chain.from_iterable(cluster["hashtags"].values))
+            sorted_most_popular_hashtags = [hashtag for hashtag, count in
+                                            Counter(aggregated_hashtags).most_common()]
```

Comment:
Czy ten `Counter` nie pochodzi z `collections`, tylko z `from typing import *`?

Reply:
W typingu nie ma Countera

Reply:
`Counter` jest z `collections`

Reply:
Jest od wersji 3.6.1: https://docs.python.org/3/library/typing.html#typing.Counter. Bardziej chodzi mi o to, że zamiast poprawnie importować `Counter` z modułu, w którym rzeczywiście się znajduje, to robimy to niejawnie poprzez `typing`.

Reply:
Dobra, to jest argument dla mnie, dlaczego `from typing import *` jest złe. Nie spotkałem się z tym wcześniej.